src/traits.rs

```diff
@@ -448,12 +444,34 @@ pub trait FindSubstring<T> {
 
 impl<'a,'b> FindSubstring<&'b [u8]> for &'a[u8] {
   fn find_substring(&self, substr: &'b[u8]) -> Option<usize> {
-    for (index,win) in self.windows(substr.len()).enumerate() {
-      if win == substr {
-        return Some(index)
+    let substr_len = substr.len();
+
+    if substr_len == 0 {
+      None
+    } else if substr_len == 1 {
+      memchr::memchr(substr[0], self)
+    } else {
+      let max = self.len() - substr_len;
+      let mut offset = 0;
+      let mut haystack = &self[..];
+
+      while let Some(position) = memchr::memchr(substr[0], haystack) {
```

Comment:
The alternative implementation would have been:
````rust
impl<'a,'b> FindSubstring<&'b [u8]> for &'a[u8] {
  fn find_substring(&self, substr: &'b[u8]) -> Option<usize> {
    let substr_len = substr.len();

    if substr_len == 0 {
      None
    } else if substr_len == 1 {
      memchr::memchr(substr[0], self)
    } else {
      let max = self.len() - substr_len;
      let haystack = &self[..max + 1];

      memchr::Memchr::new(substr[0], haystack)
          .find(|offset| &self[(*offset)..(offset+substr_len)] == substr)
    }
  }
}
````

But it doesn't pass the tests currently. Some error I don't identify yet occurs.

Not sure also how to benchmark. What do you think of it?